### PR TITLE
[IValue] Better handle sparseTensors in extractStorages

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100783

Sparse tensors don't seem to be handled when we have tensors instead
of pyobjects.

Differential Revision: [D45632427](https://our.internmc.facebook.com/intern/diff/D45632427/)

### CODE REVIEW

File: test/cpp/jit/test_misc.cpp [hunk 1]
```diff
@@ -2399,6 +2399,22 @@ TEST(FuturesTest, Basic) {
...
+// Sparse CUDA tensor test
+TEST(FutureTest, SparseTensor) {
+  // Skip test if CUDA is not available.
+  bool has_cuda = at::globalContext().hasCUDA();
+  if (!has_cuda) {
+    LOG(INFO) << "CUDA not available, skipping test";
+  }
+  auto f = c10::make_intrusive<Future>(TensorType::get());
+  at::TensorOptions opts = at::TensorOptions().device(at::DeviceType::CUDA);
+  auto sparse_tensor = at::sparse_coo_tensor(torch::ones(10), torch::ones(10), opts);
```
Comment (H-Huang) [inline test/cpp/jit/test_misc.cpp:2411]: The following should work for creating a sparse tensor:

`at::sparse_coo_tensor(torch::ones(1, 10), torch::ones(10, 10), opts);`

**Review by H-Huang — APPROVED** [review]
Just need to fix test. Stamp to unblock

File: test/cpp/jit/test_misc.cpp (hunk 1 above)
Comment (H-Huang) [reply test/cpp/jit/test_misc.cpp:2411]: Would also recommend creating a dense tensor and calling `.to_sparse()` as that will give the indices a different representation

Comment (rohan-varma) [reply test/cpp/jit/test_misc.cpp:2411]: thanks will try!

**Pull request closed without merging**